pallets/spin-polkadot/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+        // pub horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+    }
+
+    #[derive(Clone, Encode, Decode, RuntimeDebug, PartialEq, Eq, TypeInfo)]
+    pub struct GrandpaJustification<H: Header> {
+        pub round: u64,
+        pub commit: Commit<H>,
+        pub votes_ancestries: Vec<H>,
+    }
+
+    #[derive(Encode, Decode, RuntimeDebug, Clone, PartialEq, TypeInfo)]
+    pub struct AliveMessageProof<H: Header> {
+        pub fastchain_inherent_data: FastchainInherentData,
+        pub grandpa_justification: GrandpaJustification<H>,
+    }
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        type WeightInfo: crate::weights::WeightInfo;
+
+        #[pallet::constant]
+        type TimeoutBlocks: Get<BlockNumber>;
+
+        #[pallet::constant]
+        type CoolDownPeriodBlocks: Get<BlockNumber>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[derive(
+        Encode, Decode, MaxEncodedLen, TypeInfo, CloneNoBound, PartialEqNoBound, DefaultNoBound,
+    )]
+    #[scale_info(skip_type_params(T))]
+    pub enum SlowchainState<T: Config> {
+        #[default]
+        Operational {
+            last_alive_message_block_number: BlockNumberFor<T>,
+        },
+        CoolDown {
+            start_block_number: BlockNumberFor<T>,
+        },
+    }
+
+    #[pallet::storage]
+    pub type State<T: Config> = StorageValue<_, SlowchainState<T>, ValueQuery>;
+
+    // TODO: reuse pallet-staking
+    #[pallet::storage]
+    pub type ValidatorSet<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, (), ValueQuery>;
+
+    #[pallet::storage]
+    pub type AuthorityList<T: Config> =
+        StorageValue<_, sp_consensus_grandpa::AuthorityList, ValueQuery>;
+
+    #[pallet::storage]
+    pub type LastAliveMessage<T: Config> =
+        StorageValue<_, AliveMessageProof<HeaderFor<T>>, OptionQuery>;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        Heartbeat {
+            block_number: BlockNumberFor<T>,
+            who: T::AccountId,
+        },
+        StartedCoolDown {
+            block_number: BlockNumberFor<T>,
+            last_alive_message_block_number: BlockNumberFor<T>,
+        },
+        FinishedCoolDown {
+            block_number: BlockNumberFor<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        IntegerOverflow,
+        BlockTimeout,
+        CoolDownPeriod,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(current_block_number: BlockNumberFor<T>) -> Weight {
+            let mut weight = Weight::zero();
+            weight += T::DbWeight::get().reads(1);
+
+            match <State<T>>::get() {
```

Comment:
So in absence of alive messages we are in `Operational` mode for `timeout_blocks`, then in `CoolDown` for `cool_down_period_blocks` blocks and back to operational. And each alive message pushes transition to `CoolDown` for additional `timeout_blocks`. Am I understanding it correctly?

Reply:
Yes, correct

Reply:
Cool, thanks.

Reply:
By the way, could you mention this somewhere in comments or in a readme? Otherwise it is a little vague.

Reply:
Ok, I've updated pallet doc comment